File: solution/207/answer.py
class GrapNode(object):
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        grapNodes = []
        degrees = []
        for i in xrange(numCourses):
            grapNodes.append(GrapNode(i))
            degrees.append(0)
        for x in prerequisites:
            grapNodes[x[1]].neighbors.append(grapNodes[x[0]])
            degrees[x[0]] += 1
        queue = [grapNodes[i] for i in xrange(len(degrees)) if degrees[i] == 0]
        while len(queue) != 0:
            node = queue[0]
            del queue[0]
            for x in node.neighbors:
                degrees[x.label] -= 1
                if degrees[x.label] == 0:
                    queue.append(x)
        for x in degrees:
            if x:
                return False
        return True

File: solution/207/answer.cpp
struct GrapNode{
    int label;
    std::vector<GrapNode*> neighbors;
    GrapNode(int x):label(x) {};
};

class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        std::vector<GrapNode*> nodes;
        std::queue<GrapNode*> queueNodes;
        std::vector<int> degrees;
        for(int i=0;i<numCourses;i++){
            nodes.push_back(new GrapNode(i));
            degrees.push_back(0);
        }
        for(int i=0;i<prerequisites.size();i++){
            nodes[prerequisites[i].second]->neighbors.push_back(nodes[prerequisites[i].first]);
            degrees[prerequisites[i].first]++; 
        }
        for(int i=0;i<degrees.size();i++){
            if(degrees[i]==0)queueNodes.push(nodes[i]);
        }
        while(!queueNodes.empty()){
            GrapNode* node = queueNodes.front();
            queueNodes.pop();
            for(int i=0;i<node->neighbors.size();i++){
                degrees[node->neighbors[i]->label]--;
                if(degrees[node->neighbors[i]->label]==0)queueNodes.push(node->neighbors[i]);
            }
        }
        for(int i=0;i<degrees.size();i++){
            if(degrees[i]!=0)return false;
        }
        return true;
    }
};